python.mk

```diff
@@ -37,107 +40,132 @@ ${PYTHON} ${ENV_DIR} ${PIP}:
+.PHONY: coverage
+coverage: ${COVERAGE} ${ENV_DIR}/.test-requirements
+	rm -rf htmlcov/
+	${COVERAGE} erase
+	# For each setup.py file in TEST_SETUP_PY, run under coverage
+	$(foreach setup, ${COVERAGE_SETUP_PY}, \
+		$(if ${TEST_PYTHON_PATH},PYTHONPATH=${TEST_PYTHON_PATH},) ${COMPILER_FLAGS} ${COVERAGE} run $(if ${COVERAGE_SRC_DIR},--source ${COVERAGE_SRC_DIR},) $(if ${COVERAGE_EXCL},--omit "${COVERAGE_EXCL}",) -a ${setup} test &&) true
+	${COVERAGE} combine
+	${COVERAGE} report -m --fail-under 100
+	${COVERAGE} html
+
+# Dummy test requirements target to which we add the requirements for each component,
+# so that coverage can depend on this target.
+${ENV_DIR}/.test-requirements:
 	touch $@
 
-# Common rules to build a python component
+# Common rules for a python component that includes tests
+# @param $1 - Target name
+#
+# Each target must supply:
+#   - <target>_TEST_REQUIREMENTS - A list of the requirements files that the target uses
+#   - <target>_TEST_SETUP        - The setup.py file used to run the tests
+define python_test_component
+
+${ENV_DIR}/.$1-test-requirements: $${$1_TEST_REQUIREMENTS} ${ENV_DIR}/.env
+	# Install the test requirements for this component
+	$$(foreach reqs, $${$1_TEST_REQUIREMENTS}, ${PIP} install -r $${reqs} &&) true
+	touch $$@
+
+.PHONY: $1_test
+$1_test: ${ENV_DIR}/.$1-test-requirements
+	$(if ${TEST_PYTHON_PATH},PYTHONPATH=${TEST_PYTHON_PATH},) ${COMPILER_FLAGS} ${PYTHON} $${$1_TEST_SETUP} test -v
+
+${ENV_DIR}/.test-requirements: ${ENV_DIR}/.$1-test-requirements
+
+test: $1_test
+endef
+
+# Common rules to build a python component. Includes adding a test target.
 #
 # @param $1 - Target name
 #
 # Each target must supply:
 #   - <target>_SETUP        - A list of the setup.py files that the target depends on
 #   - <target>_REQUIREMENTS - A list of the requirements files that the target uses
 #   - <target>_SOURCES      - A list of the python source files that go into the wheel
-#   - <target>_WHEELS       - A list of the wheels that are produced
+#   - <target>_WHEELS       - (Optional) A list of built wheels that should be installed in the environent.
 #   - <target>_FLAGS        - (Optional) Extra flags to be passed to python commands
+#   - <target>_BUILD_DIRS   - (Optional) If specified, this component uses a the build directory {ROOT}/build_<target>
 #
 # The location of the wheelhouse can be overridden if desired, by specifying
 # <target>_WHEELHOUSE
 define python_component
 
+# Create common tests targets for this component
+$(call python_test_component,$1)
+
 # The wheelhouse can be overridden if desired
 $1_WHEELHOUSE ?= $1_wheelhouse
 
-${ENV_DIR}/.wheels-cleaned: ${ENV_DIR}/.$1-clean-wheels
-${ENV_DIR}/.wheels-built: ${ENV_DIR}/.$1-build-wheels
-${ENV_DIR}/.wheels-installed: ${ENV_DIR}/.$1-install-wheels
+# Whenever the requirements change, we must delete our venv as we may have the
+# wrong requirements installed
+${ENV_DIR}/.env: $${$1_REQUIREMENTS}
```

Comment:
Should we also do this for test requirements?

Reply:
Yep, now done